homeassistant/components/homeassistant_hardware/silabs_multiprotocol_addon.py

```diff
@@ -584,23 +683,205 @@ async def async_step_notify_channel_change(
     async def async_step_uninstall_addon(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
-        """Uninstall the addon (not implemented)."""
-        return await self.async_step_show_revert_guide()
+        """Uninstall the addon and revert the firmware."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="uninstall_addon",
+                data_schema=vol.Schema(
+                    {vol.Required(CONF_DISABLE_MULTI_PAN, default=False): bool}
+                ),
+                description_placeholders={"hardware_name": self._hardware_name()},
+            )
+        if not user_input[CONF_DISABLE_MULTI_PAN]:
+            return self.async_create_entry(title="", data={})
+
+        return await self.async_step_firmware_revert()
 
-    async def async_step_show_revert_guide(
+    async def async_step_firmware_revert(
         self, user_input: dict[str, Any] | None = None
     ) -> FlowResult:
-        """Link to a guide for reverting to Zigbee firmware."""
-        if user_input is None:
-            return self.async_show_form(step_id="show_revert_guide")
-        return self.async_create_entry(title="", data={})
+        """Install the flasher addon, if necessary."""
 
-    async def async_step_addon_installed_other_device(
+        addon_manager: AddonManager = get_flasher_addon_manager(self.hass)
+        addon_info = await self._async_get_addon_info(addon_manager)
+
+        if addon_info.state == AddonState.NOT_INSTALLED:
+            return await self.async_step_install_flasher_addon()
+
+        if addon_info.state == AddonState.NOT_RUNNING:
+            return await self.async_step_configure_flasher_addon()
+
+        # If the addon is already installed and running, fail
+        return self.async_abort(
+            reason="addon_install_failed",
```

Comment:
Could we give a better error message? This means flashing is actually ongoing (or is hung)?